### PR TITLE
SLING-12019 - Avoid duplicate ResourceResolverFactory registrations

### PR DESCRIPTION
**Pull request by jsedding**
- reworked synchronization from relying on worker queue to locks
- increased test coverage, particularly for concurrent scenarios
- switched to junit-platform to allow mixing junit4 and junit5 tests
- enabled parallel test execution for junit-platform

### CODE REVIEW

File: src/main/java/org/apache/sling/resourceresolver/impl/FactoryRegistrationHandler.java [hunk 1]
```diff
@@ -30,45 +30,85 @@
...
+            LOG.debug("activator differs = {}, factoryPreconditions differ = {}", this.activator != activator, !Objects.equals(this.factoryPreconditions, factoryPreconditions));
+            LOG.debug("factoryPreconditions {} vs {}", this.factoryPreconditions, factoryPreconditions);
+            this.factoryPreconditions = factoryPreconditions;
+            this.activator = activator;
+        } finally {
+            configurationLock.unlock();
+        }
+
+        if (reRegister) {
+            if (this.factoryRegistration != null) {
```
Comment (rombert) [inline src/main/java/org/apache/sling/resourceresolver/impl/FactoryRegistrationHandler.java:88]: Should this check be done inside the critical region protected by the `configurationLock`?

File: src/main/java/org/apache/sling/resourceresolver/impl/FactoryRegistrationHandler.java [hunk 2]
```diff
@@ -77,15 +117,24 @@ public void close() {
...
+            try {
+                configurationLock.lock();
+                preconditionsOk = factoryPreconditions.checkPreconditions(unavailableName, unavailableServicePid);
+                localActivator = this.activator;
+            } finally {
+                configurationLock.unlock();
+            }
+
+            factoryRegistrationWorker.execute(() -> {
+                if (preconditionsOk && this.factoryRegistration == null) {
```
Comment (rombert) [inline src/main/java/org/apache/sling/resourceresolver/impl/FactoryRegistrationHandler.java:131]: What are the consequences of making an "unsychronized" read of `this.factoryRegistration` in this method?

**Review by rombert — COMMENTED** [review]
I haven't grokked this fully, but I least as I can ask some questions :-)

File: src/main/java/org/apache/sling/resourceresolver/impl/FactoryRegistrationHandler.java (hunk 1 above)
Comment (jsedding) [reply src/main/java/org/apache/sling/resourceresolver/impl/FactoryRegistrationHandler.java:88]: You are right to spot the missing lock. But the `registrationLock` is required when accessing `factoryRegistration`.

File: src/main/java/org/apache/sling/resourceresolver/impl/FactoryRegistrationHandler.java (hunk 2 above)
Comment (jsedding) [reply src/main/java/org/apache/sling/resourceresolver/impl/FactoryRegistrationHandler.java:131]: Again, same as abov. I'll add the `registrationLock` here.

**Comment by jsedding** [comment]
Thanks @rombert! While fixing the missing locks, I realized that many of the null checks on `factoryRegistration` were unnecessary. And once I had removed them, it seemed clearer to use an `AtomicReference` for the `factoryRegistration` instead of the locks.

**Review by rombert — APPROVED** [review]

**Pull request merged**